WIP:docker: use alpine flavor of postgres image

leonklingele opened the pull request:
Image size reduced by over 68%, from `230MB` down to `72.4MB`.
Additionally, this seems to free around 2GB of intermediary images.

leonklingele commented:
No database upgrade happened on my end and everything works fine after migrating to the new image, but please test this more carefully.

leonklingele commented:
Can someone already running a Docker cluster please test this change (make sure to backup your postgres database first). I noticed this strange behavior on one of my instances (might be unrelated to this change though):

<img width="1325" alt="Bildschirmfoto 2019-08-10 um 14 43 02" src="https://user-images.githubusercontent.com/5585491/62884670-66f9ab00-bd37-11e9-94fa-25667f8f85f3.png">

@omarroth any idea on what might cause this?

omarroth commented:
My only guess would be that duplicate videos are being insert into the database. Would you mind checking that `videos` has `id` set as a primary key? Something like:
```
invidious=# \d+ videos
...
Indexes:
    "videos_pkey" PRIMARY KEY, btree (id)
    "videos_id_idx" UNIQUE, btree (id)
```

leonklingele commented:
Unfortunately, I have removed the "broken" database since and this never happened again.

postgres was logging lots of `ERROR:  duplicate key value violates unique constraint "videos_pkey"` messages:

```
2019-08-10 12:54:47.961 UTC [549] ERROR:  duplicate key value violates unique constraint "videos_pkey"
2019-08-10 12:54:47.961 UTC [549] DETAIL:  Key (id)=(REDACTED) already exists.
2019-08-10 12:54:47.961 UTC [549] STATEMENT:  UPDATE videos SET (info,updated,title,views,likes,dislikes,wilson_score,published,description,language,author,ucid,allowed_regions,is_family_friendly,genre,genre_url,license,sub_count_text,author_thumbnail)= (($2),($3),($4),($5),($6),($7),($8),($9),($10),($11),($12),($13),($14),($15),($16),($17),($18),($19),($20)) WHERE id = $1
````

How can an `UPDATE` even result in duplicate keys if the keyed-field is not touched?

omarroth commented:
> How can an `UPDATE` even result in duplicate keys if the keyed-field is not touched?

If there were already duplicates in the database, then an `UPDATE` would be performing an operation on multiple rows. I'm not sure how that would happen though, since if that constraint were dropped it shouldn't be able to be created again with duplicates still in the DB.

omarroth commented:
Is this still WIP? From some quick testing it LGTM.

leonklingele commented:
It would be nice if someone could test this change when "migrating" from an existing postgres Docker volume. See https://github.com/omarroth/invidious/pull/687#issuecomment-520519119 for what problem(s) might occur.
I'll do some more testing later this week and will report back to you.

leonklingele commented:
I'm still not sure / haven't tried to reproduce whether this change is responsible for spoiling the database as described in https://github.com/omarroth/invidious/pull/687#issuecomment-520519119.

Anyone, feel free to test this with a backup of your instance.

SISheogorath commented:
Be careful with postgres and the alpine base image. I had to learn the hard way that the debian based vs the alpine based images are not data-compatible.

https://github.com/docker-library/postgres/issues/276

This can be solved by dumping and re-importing the db but this needs to be documented in upgrade instructions

tleydxdy commented on docker/entrypoint.postgres.sh:
```diff
@@ -5,7 +5,7 @@ if [ ! -f /var/lib/postgresql/data/setupFinished ]; then
     /usr/local/bin/docker-entrypoint.sh postgres &
     sleep 10
-    until runuser -l postgres -c 'pg_isready' 2>/dev/null; do
+    until su postgres -c 'pg_isready' 2>/dev/null; do
```
I think this needs to be `su postgres -s /bin/sh -c 'pg_isready'` same for the rest of the `su`s because the default shell for postgres user is nologin on the alpine system.

tleydxdy replied:
The database setup fails when I try to do use the patch as is

leonklingele replied:
@tleydxdy works fine for me as-is. What error do you get?

tleydxdy replied:
it gives a error: "This account is not available"

tleydxdy replied:
are you trying it with a database existing already? if so it wouldn't get to the su parts

tleydxdy replied:
Here are the logs with a clean install
```
### first run - setting up invidious database
****************************************************
WARNING: No password has been set for the database.
         This will allow anyone with access to the
         Postgres port to access your database. In
         Docker's default configuration, this is
         effectively any other container on the same
         system.

         Use "-e POSTGRES_PASSWORD=password" to set
         it in "docker run".
****************************************************
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default timezone ... UTC
selecting dynamic shared memory implementation ... posix
creating configuration files ... ok
running bootstrap script ... ok
sh: locale: not found
2020-03-19 23:53:09.838 UTC [31] WARNING:  no usable system locales were found
performing post-bootstrap initialization ... ok
syncing data to disk ... This account is not available
ok

WARNING: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    pg_ctl -D /var/lib/postgresql/data -l logfile start

waiting for server to start....2020-03-19 23:53:22.135 UTC [39] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2020-03-19 23:53:22.192 UTC [40] LOG:  database system was shut down at 2020-03-19 23:53:12 UTC
2020-03-19 23:53:22.203 UTC [39] LOG:  database system is ready to accept connections
 done
server started

/usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*

2020-03-19 23:53:22.222 UTC [39] LOG:  received fast shutdown request
waiting for server to shut down....2020-03-19 23:53:22.232 UTC [39] LOG:  aborting any active transactions
2020-03-19 23:53:22.233 UTC [39] LOG:  worker process: logical replication launcher (PID 46) exited with exit code 1
2020-03-19 23:53:22.234 UTC [41] LOG:  shutting down
2020-03-19 23:53:22.312 UTC [39] LOG:  database system is shut down
 done
server stopped

PostgreSQL init process complete; ready for start up.

2020-03-19 23:53:22.348 UTC [6] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2020-03-19 23:53:22.348 UTC [6] LOG:  listening on IPv6 address "::", port 5432
2020-03-19 23:53:22.363 UTC [6] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2020-03-19 23:53:22.404 UTC [48] LOG:  database system was shut down at 2020-03-19 23:53:22 UTC
2020-03-19 23:53:22.415 UTC [6] LOG:  database system is ready to accept connections
### Postgres is unavailable - waiting
2020-03-19 23:53:24.485 UTC [56] FATAL:  role "kemal" does not exist
2020-03-19 23:53:28.297 UTC [57] FATAL:  role "kemal" does not exist
This account is not available
### Postgres is unavailable - waiting
2020-03-19 23:53:33.453 UTC [62] FATAL:  role "kemal" does not exist
This account is not available
2020-03-19 23:53:41.830 UTC [72] FATAL:  role "kemal" does not exist
### Postgres is unavailable - waiting
This account is not available
### Postgres is unavailable - waiting
2020-03-19 23:53:56.540 UTC [78] FATAL:  role "kemal" does not exist
This account is not available
```

leonklingele replied:
Hmm, strange! Works fine for me:

```sh
$ cd $(mktemp -d)
$ git clone https://github.com/omarroth/invidious
$ cd invidious/
$ git fetch origin pull/687/head:docker-postgres-alpine
$ git checkout docker-postgres-alpine
$ docker-compose up --build
Creating network "invidious_default" with the default driver
Creating volume "invidious_postgresdata" with default driver
Building postgres
Step 1/7 : FROM postgres:10-alpine
10-alpine: Pulling from library/postgres
c9b1b535fdd9: Pull complete
d8f3047c2e42: Pull complete
f2e53fddf183: Pull complete
9c8282d512e5: Pull complete
65cbae707289: Pull complete
bafd14c3c0dc: Pull complete
ea597e72d12f: Pull complete
3fc257549592: Pull complete
0bd4d7958341: Pull complete
Digest: sha256:c82ed22ac2c571d1cb17f1773018527921a977745684b786619b7bd2e11c4960
Status: Downloaded newer image for postgres:10-alpine
 ---> 97f1045583bf
Step 2/7 : ENV POSTGRES_USER postgres
 ---> Running in 75aec12e9ce4
Removing intermediate container 75aec12e9ce4
 ---> 8a36ac3ff04b
Step 3/7 : ENV POSTGRES_HOST_AUTH_METHOD trust
 ---> Running in 36eb53b09aa6
Removing intermediate container 36eb53b09aa6
 ---> 0e95f9c52424
Step 4/7 : ADD ./config/sql /config/sql
 ---> afc78960f7c6
Step 5/7 : ADD ./docker/entrypoint.postgres.sh /entrypoint.sh
 ---> c1db8ec47953
Step 6/7 : ENTRYPOINT [ "/entrypoint.sh" ]
 ---> Running in c13b9172e37b
Removing intermediate container c13b9172e37b
 ---> da15dc762d75
Step 7/7 : CMD [ "postgres" ]
 ---> Running in 0abeaa862e86
Removing intermediate container 0abeaa862e86
 ---> 3c862a911d1a
Successfully built 3c862a911d1a
Successfully tagged invidious_postgres:latest
Building invidious
Step 1/22 : FROM alpine:edge AS builder
edge: Pulling from library/alpine
2384fa44dfc2: Pull complete
Digest: sha256:2849e06886c579433ccf3738e36017dc2b354a91d640fb01761d5b70f98d0e05
Status: Downloaded newer image for alpine:edge
 ---> 24cae4d038c0
Step 2/22 : RUN apk add --no-cache crystal shards libc-dev     yaml-dev libxml2-dev sqlite-dev zlib-dev openssl-dev     sqlite-static zlib-static openssl-libs-static
 ---> Running in 612d25e3f184
fetch http://dl-cdn.alpinelinux.org/alpine/edge/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz
(1/58) Upgrading musl (1.1.24-r2 -> 1.1.24-r3)
(2/58) Installing gc (8.0.4-r1)
(3/58) Installing libgcc (9.2.0-r6)
(4/58) Installing libstdc++ (9.2.0-r6)
(5/58) Installing libgc++ (8.0.4-r1)
(6/58) Installing pkgconf (1.6.3-r2)
(7/58) Installing gc-dev (8.0.4-r1)
(8/58) Installing libatomic_ops (7.6.10-r1)
(9/58) Installing libbz2 (1.0.8-r1)
(10/58) Installing expat (2.2.9-r1)
(11/58) Installing libffi (3.2.1-r6)
(12/58) Installing gdbm (1.13-r1)
(13/58) Installing xz-libs (5.2.5-r0)
(14/58) Installing ncurses-terminfo-base (6.2_p20200308-r0)
(15/58) Installing ncurses-libs (6.2_p20200308-r0)
(16/58) Installing readline (8.0.4-r0)
(17/58) Installing sqlite-libs (3.31.1-r1)
(18/58) Installing python3 (3.8.2-r0)
(19/58) Installing libevent (2.1.11-r0)
(20/58) Installing libevent-dev (2.1.11-r0)
(21/58) Installing libevent-static (2.1.11-r0)
(22/58) Installing binutils (2.33.1-r0)
(23/58) Installing gmp (6.2.0-r0)
(24/58) Installing isl (0.18-r0)
(25/58) Installing libgomp (9.2.0-r6)
(26/58) Installing libatomic (9.2.0-r6)
(27/58) Installing libgphobos (9.2.0-r6)
(28/58) Installing mpfr4 (4.0.2-r1)
(29/58) Installing mpc1 (1.1.0-r1)
(30/58) Installing gcc (9.2.0-r6)
(31/58) Installing libgmpxx (6.2.0-r0)
(32/58) Installing gmp-dev (6.2.0-r0)
(33/58) Installing libpcre16 (8.44-r0)
(34/58) Installing libpcre32 (8.44-r0)
(35/58) Installing pcre (8.44-r0)
(36/58) Installing libpcrecpp (8.44-r0)
(37/58) Installing pcre-dev (8.44-r0)
(38/58) Installing llvm5-libs (5.0.2-r3)
(39/58) Installing crystal (0.31.1-r2)
(40/58) Installing musl-dev (1.1.24-r3)
(41/58) Installing libc-dev (0.7.2-r1)
(42/58) Installing zlib-dev (1.2.11-r3)
(43/58) Installing xz-dev (5.2.5-r0)
(44/58) Installing libxml2 (2.9.10-r3)
(45/58) Installing libxml2-dev (2.9.10-r3)
(46/58) Installing openssl-dev (1.1.1d-r5)
(47/58) Installing openssl-libs-static (1.1.1d-r5)
(48/58) Installing ca-certificates (20191127-r2)
(49/58) Installing nghttp2-libs (1.40.0-r0)
(50/58) Installing libcurl (7.69.1-r0)
(51/58) Installing pcre2 (10.34-r1)
(52/58) Installing git (2.25.2-r0)
(53/58) Installing yaml (0.2.2-r1)
(54/58) Installing shards (0.9.0-r1)
(55/58) Installing sqlite-dev (3.31.1-r1)
(56/58) Installing sqlite-static (3.31.1-r1)
(57/58) Installing yaml-dev (0.2.2-r1)
(58/58) Installing zlib-static (1.2.11-r3)
Executing busybox-1.31.1-r13.trigger
Executing ca-certificates-20191127-r2.trigger
OK: 298 MiB in 71 packages
Removing intermediate container 612d25e3f184
 ---> 946af048c8d1
Step 3/22 : WORKDIR /invidious
 ---> Running in 9cc1c76615a0
Removing intermediate container 9cc1c76615a0
 ---> 3b5149a74898
Step 4/22 : COPY ./shard.yml ./shard.yml
 ---> d6e53093a9c3
Step 5/22 : RUN shards update && shards install
 ---> Running in a510eae24a8f
Resolving dependencies
Fetching https://github.com/will/crystal-pg.git
Fetching https://github.com/crystal-lang/crystal-db.git
Fetching https://github.com/crystal-lang/crystal-sqlite3.git
Fetching https://github.com/kemalcr/kemal.git
Fetching https://github.com/luislavena/radix.git
Fetching https://github.com/jeromegn/kilt.git
Fetching https://github.com/crystal-loot/exception_page.git
Fetching https://github.com/ysbaddaden/pool.git
Fetching https://github.com/omarroth/protodec.git
Fetching https://github.com/omarroth/lsquic.cr.git
Installing pg (0.19.0)
Installing sqlite3 (0.14.0)
Installing kemal (0.26.1)
Installing pool (0.2.3)
Installing protodec (0.1.2)
Installing lsquic (0.1.9)
Installing radix (0.3.9)
Installing exception_page (0.1.4)
Installing db (0.7.0)
Installing kilt (0.4.0)
Writing shard.lock
Resolving dependencies
Fetching https://github.com/will/crystal-pg.git
Fetching https://github.com/crystal-lang/crystal-db.git
Fetching https://github.com/crystal-lang/crystal-sqlite3.git
Fetching https://github.com/kemalcr/kemal.git
Fetching https://github.com/luislavena/radix.git
Fetching https://github.com/jeromegn/kilt.git
Fetching https://github.com/crystal-loot/exception_page.git
Fetching https://github.com/ysbaddaden/pool.git
Fetching https://github.com/omarroth/protodec.git
Fetching https://github.com/omarroth/lsquic.cr.git
Using pg (0.19.0)
Using sqlite3 (0.14.0)
Using kemal (0.26.1)
Using pool (0.2.3)
Using protodec (0.1.2)
Using lsquic (0.1.9)
Using radix (0.3.9)
Using exception_page (0.1.4)
Using db (0.7.0)
Using kilt (0.4.0)
Writing shard.lock
Removing intermediate container a510eae24a8f
 ---> c9620ac663c6
Step 6/22 : RUN apk add --no-cache curl &&     curl -Lo /etc/apk/keys/omarroth.rsa.pub https://github.com/omarroth/boringssl-alpine/releases/download/1.1.0-r0/omarroth.rsa.pub &&     curl -Lo boringssl-dev.apk https://github.com/omarroth/boringssl-alpine/releases/download/1.1.0-r0/boringssl-dev-1.1.0-r0.apk &&     curl -Lo lsquic.apk https://github.com/omarroth/lsquic-alpine/releases/download/2.6.3-r0/lsquic-2.6.3-r0.apk &&     tar -xf boringssl-dev.apk &&     tar -xf lsquic.apk
 ---> Running in 463ddadcb915
fetch http://dl-cdn.alpinelinux.org/alpine/edge/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz
(1/1) Installing curl (7.69.1-r0)
Executing busybox-1.31.1-r13.trigger
OK: 299 MiB in 72 packages
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   607  100   607    0     0   1873      0 --:--:-- --:--:-- --:--:--  1867
100   451  100   451    0     0    561      0 --:--:-- --:--:-- --:--:--   561
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   632  100   632    0     0   2012      0 --:--:-- --:--:-- --:--:--  2006
100 9958k  100 9958k    0     0  2608k      0  0:00:03  0:00:03 --:--:-- 3731k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   625  100   625    0     0   1693      0 --:--:-- --:--:-- --:--:--  1693
100  673k  100  673k    0     0   477k      0  0:00:01  0:00:01 --:--:--  662k
Removing intermediate container 463ddadcb915
 ---> cc571a3b6785
Step 7/22 : RUN mv ./usr/lib/libcrypto.a ./lib/lsquic/src/lsquic/ext/libcrypto.a &&     mv ./usr/lib/libssl.a ./lib/lsquic/src/lsquic/ext/libssl.a &&     mv ./usr/lib/liblsquic.a ./lib/lsquic/src/lsquic/ext/liblsquic.a
 ---> Running in 62f4151a0070
Removing intermediate container 62f4151a0070
 ---> e274fcf2e6f3
Step 8/22 : COPY ./src/ ./src/
 ---> 47ff331ff008
Step 9/22 : COPY ./.git/ ./.git/
 ---> 59d7be2a095b
Step 10/22 : RUN crystal build ./src/invidious.cr     --static --warnings all --error-on-warnings     -Dmusl     --link-flags "-lxml2 -llzma"
 ---> Running in 3f4daa0e3399
Removing intermediate container 3f4daa0e3399
 ---> 38bb3f975617
Step 11/22 : FROM alpine:latest
latest: Pulling from library/alpine
c9b1b535fdd9: Already exists
Digest: sha256:ab00606a42621fb68f2ed6ad3c88be54397f981a7b70a79db3d1172b11c4367d
Status: Downloaded newer image for alpine:latest
 ---> e7d92cdc71fe
Step 12/22 : RUN apk add --no-cache librsvg ttf-opensans
 ---> Running in de1dce3dc2d5
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
(1/39) Installing libxau (1.0.9-r0)
(2/39) Installing libbsd (0.10.0-r0)
(3/39) Installing libxdmcp (1.1.3-r0)
(4/39) Installing libxcb (1.13.1-r0)
(5/39) Installing libx11 (1.6.9-r0)
(6/39) Installing libxext (1.3.4-r0)
(7/39) Installing libxrender (0.9.10-r3)
(8/39) Installing expat (2.2.9-r1)
(9/39) Installing libbz2 (1.0.8-r1)
(10/39) Installing libpng (1.6.37-r1)
(11/39) Installing freetype (2.10.1-r0)
(12/39) Installing libuuid (2.34-r1)
(13/39) Installing fontconfig (2.13.1-r2)
(14/39) Installing pixman (0.38.4-r0)
(15/39) Installing cairo (1.16.0-r2)
(16/39) Installing libffi (3.2.1-r6)
(17/39) Installing libintl (0.20.1-r2)
(18/39) Installing libblkid (2.34-r1)
(19/39) Installing libmount (2.34-r1)
(20/39) Installing pcre (8.43-r0)
(21/39) Installing glib (2.62.6-r0)
(22/39) Installing xz-libs (5.2.4-r0)
(23/39) Installing libxml2 (2.9.10-r2)
(24/39) Installing libcroco (0.6.13-r1)
(25/39) Installing libgcc (9.2.0-r4)
(26/39) Installing pkgconf (1.6.3-r0)
(27/39) Installing shared-mime-info (1.15-r0)
(28/39) Installing libjpeg-turbo (2.0.4-r0)
(29/39) Installing tiff (4.1.0-r0)
(30/39) Installing gdk-pixbuf (2.40.0-r0)
(31/39) Installing libxft (2.3.3-r0)
(32/39) Installing fribidi (1.0.8-r0)
(33/39) Installing graphite2 (1.3.13-r1)
(34/39) Installing harfbuzz (2.6.4-r0)
(35/39) Installing pango (1.44.7-r0)
(36/39) Installing librsvg (2.46.4-r0)
(37/39) Installing libfontenc (1.1.4-r0)
(38/39) Installing mkfontscale (1.2.1-r1)
(39/39) Installing ttf-opensans (1.10-r0)
Executing ttf-opensans-1.10-r0.post-install
Executing busybox-1.31.1-r9.trigger
Executing fontconfig-2.13.1-r2.trigger
Executing shared-mime-info-1.15-r0.trigger
Executing gdk-pixbuf-2.40.0-r0.trigger
Executing mkfontscale-1.2.1-r1.trigger
OK: 32 MiB in 53 packages
Removing intermediate container de1dce3dc2d5
 ---> 124cdc0a3fb5
Step 13/22 : WORKDIR /invidious
 ---> Running in e0977b92a95f
Removing intermediate container e0977b92a95f
 ---> 9ee75f4c8fc1
Step 14/22 : RUN addgroup -g 1000 -S invidious &&     adduser -u 1000 -S invidious -G invidious
 ---> Running in 46153d75fd68
Removing intermediate container 46153d75fd68
 ---> 621124352e50
Step 15/22 : COPY ./assets/ ./assets/
 ---> 26bf0b855a8e
Step 16/22 : COPY ./config/config.yml ./config/config.yml
 ---> 7caaa60dcdd9
Step 17/22 : COPY ./config/sql/ ./config/sql/
 ---> 637edb50b591
Step 18/22 : COPY ./locales/ ./locales/
 ---> 45caa3c8acb0
Step 19/22 : RUN sed -i 's/host: \(127.0.0.1\|localhost\)/host: postgres/' config/config.yml
 ---> Running in 1a245c36d6ce
Removing intermediate container 1a245c36d6ce
 ---> b87e4b19fc91
Step 20/22 : COPY --from=builder /invidious/invidious .
 ---> bd649d74f07e
Step 21/22 : USER invidious
 ---> Running in 0ec295379e71
Removing intermediate container 0ec295379e71
 ---> 0af3da62260c
Step 22/22 : CMD [ "/invidious/invidious" ]
 ---> Running in 301959e2bb5e
Removing intermediate container 301959e2bb5e
 ---> 7c14ec8da4ad
Successfully built 7c14ec8da4ad
Successfully tagged invidious_invidious:latest
Creating invidious_postgres_1 ... done
Creating invidious_invidious_1 ... done
Attaching to invidious_postgres_1, invidious_invidious_1
postgres_1   | ### first run - setting up invidious database
postgres_1   | ********************************************************************************
postgres_1   | WARNING: POSTGRES_HOST_AUTH_METHOD has been set to "trust". This will allow
postgres_1   |          anyone with access to the Postgres port to access your database without
postgres_1   |          a password, even if POSTGRES_PASSWORD is set. See PostgreSQL
postgres_1   |          documentation about "trust":
postgres_1   |          https://www.postgresql.org/docs/current/auth-trust.html
postgres_1   |          In Docker's default configuration, this is effectively any other
postgres_1   |          container on the same system.
postgres_1   |
postgres_1   |          It is not recommended to use POSTGRES_HOST_AUTH_METHOD=trust. Replace
postgres_1   |          it with "-e POSTGRES_PASSWORD=password" instead to set a password in
postgres_1   |          "docker run".
postgres_1   | ********************************************************************************
postgres_1   | The files belonging to this database system will be owned by user "postgres".
postgres_1   | This user must also own the server process.
postgres_1   |
postgres_1   | The database cluster will be initialized with locale "en_US.utf8".
postgres_1   | The default database encoding has accordingly been set to "UTF8".
postgres_1   | The default text search configuration will be set to "english".
postgres_1   |
postgres_1   | Data page checksums are disabled.
postgres_1   |
postgres_1   | fixing permissions on existing directory /var/lib/postgresql/data ... ok
postgres_1   | creating subdirectories ... ok
postgres_1   | selecting default max_connections ... 100
postgres_1   | selecting default shared_buffers ... 128MB
postgres_1   | selecting default timezone ... UTC
postgres_1   | selecting dynamic shared memory implementation ... posix
postgres_1   | creating configuration files ... ok
postgres_1   | running bootstrap script ... ok
postgres_1   | performing post-bootstrap initialization ... sh: locale: not found
postgres_1   | 2020-03-21 22:45:49.575 UTC [32] WARNING:  no usable system locales were found
postgres_1   | ok
postgres_1   | syncing data to disk ... ok
postgres_1   |
postgres_1   | Success. You can now start the database server using:
postgres_1   |
postgres_1   |     pg_ctl -D /var/lib/postgresql/data -l logfile start
postgres_1   |
postgres_1   |
postgres_1   | WARNING: enabling "trust" authentication for local connections
postgres_1   | You can change this by editing pg_hba.conf or using the option -A, or
postgres_1   | --auth-local and --auth-host, the next time you run initdb.
postgres_1   | waiting for server to start....2020-03-21 22:45:50.182 UTC [37] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
postgres_1   | 2020-03-21 22:45:50.192 UTC [38] LOG:  database system was shut down at 2020-03-21 22:45:49 UTC
postgres_1   | 2020-03-21 22:45:50.195 UTC [37] LOG:  database system is ready to accept connections
postgres_1   |  done
postgres_1   | server started
postgres_1   |
postgres_1   | /usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*
postgres_1   |
postgres_1   | waiting for server to shut down....2020-03-21 22:45:50.277 UTC [37] LOG:  received fast shutdown request
postgres_1   | 2020-03-21 22:45:50.278 UTC [37] LOG:  aborting any active transactions
postgres_1   | 2020-03-21 22:45:50.279 UTC [37] LOG:  worker process: logical replication launcher (PID 44) exited with exit code 1
postgres_1   | 2020-03-21 22:45:50.279 UTC [39] LOG:  shutting down
postgres_1   | 2020-03-21 22:45:50.288 UTC [37] LOG:  database system is shut down
postgres_1   |  done
postgres_1   | server stopped
postgres_1   |
postgres_1   | PostgreSQL init process complete; ready for start up.
postgres_1   |
postgres_1   | 2020-03-21 22:45:50.385 UTC [6] LOG:  listening on IPv4 address "0.0.0.0", port 5432
postgres_1   | 2020-03-21 22:45:50.385 UTC [6] LOG:  listening on IPv6 address "::", port 5432
postgres_1   | 2020-03-21 22:45:50.387 UTC [6] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
postgres_1   | 2020-03-21 22:45:50.397 UTC [46] LOG:  database system was shut down at 2020-03-21 22:45:50 UTC
invidious_1  | Unhandled exception:  (DB::ConnectionRefused)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  | Caused by: Cannot establish connection (PQ::ConnectionError)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  | Caused by: Error connecting to 'postgres:5432': Connection refused (Errno)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
postgres_1   | 2020-03-21 22:45:50.399 UTC [6] LOG:  database system is ready to accept connections
invidious_1  | Unhandled exception:  (DB::ConnectionRefused)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  | Caused by: role "kemal" does not exist (PQ::PQError)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_invidious_1 exited with code 1
postgres_1   | 2020-03-21 22:45:52.663 UTC [53] FATAL:  role "kemal" does not exist
invidious_1  | Unhandled exception:  (DB::ConnectionRefused)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  | Caused by: role "kemal" does not exist (PQ::PQError)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_invidious_1 exited with code 1
postgres_1   | 2020-03-21 22:45:55.588 UTC [54] FATAL:  role "kemal" does not exist
postgres_1   | 2020-03-21 22:45:58.663 UTC [55] FATAL:  role "kemal" does not exist
postgres_1   | /var/run/postgresql:5432 - accepting connections
postgres_1   | ### importing table schemas
postgres_1   | CREATE ROLE
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE INDEX
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE TYPE
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | CREATE TABLE
postgres_1   | GRANT
postgres_1   | ### invidious database setup finished
postgres_1   |
postgres_1   | PostgreSQL Database directory appears to contain a database; Skipping initialization
postgres_1   |
postgres_1   | 2020-03-21 22:46:00.232 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
postgres_1   | 2020-03-21 22:46:00.233 UTC [1] LOG:  listening on IPv6 address "::", port 5432
postgres_1   | 2020-03-21 22:46:00.235 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
postgres_1   | 2020-03-21 22:46:00.245 UTC [21] LOG:  database system was interrupted; last known up at 2020-03-21 22:45:59 UTC
postgres_1   | 2020-03-21 22:46:00.399 UTC [21] LOG:  database system was not properly shut down; automatic recovery in progress
postgres_1   | 2020-03-21 22:46:00.402 UTC [21] LOG:  redo starts at 0/163E370
postgres_1   | 2020-03-21 22:46:00.407 UTC [21] LOG:  invalid record length at 0/169F760: wanted 24, got 0
postgres_1   | 2020-03-21 22:46:00.407 UTC [21] LOG:  redo done at 0/169F6F8
postgres_1   | 2020-03-21 22:46:00.407 UTC [21] LOG:  last completed transaction was at log time 2020-03-21 22:45:59.574403+00
invidious_1  | Unhandled exception:  (DB::ConnectionRefused)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  | Caused by: role "kemal" does not exist (PQ::PQError)
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_1  |   from ???
invidious_invidious_1 exited with code 1

$ curl -vs $(docker-machine ip):3000 | head -n5
*   Trying 192.168.99.135:3000...
* TCP_NODELAY set
* Connected to 192.168.99.135 (192.168.99.135) port 3000 (#0)
> GET / HTTP/1.1
> Host: 192.168.99.135:3000
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Connection: keep-alive
< Content-Type: text/html
< X-Frame-Options: sameorigin
< X-XSS-Protection: 1; mode=block
< X-Content-Type-Options: nosniff
< Content-Security-Policy: default-src 'none'; script-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self' data:; connect-src 'self'; media-src 'self' blob: https://*.googlevideo.com:443
< Referrer-Policy: same-origin
< Content-Length: 6203
<
{ [6203 bytes data]
* Connection #0 to host 192.168.99.135 left intact
<!DOCTYPE html>
<html lang="en-US">

<head>
    <meta charset="utf-8">
```

tleydxdy replied:
err! why is my setup always different than others :( I'm using Debian 10 with latest updates and the provided docker, yet only I seems to have these strange problems

Perflyst replied:
because probably nobody uses docker from debian but https://docs.docker.com/install/linux/docker-ce/debian/ instead

tleydxdy replied:
a.k.a "docker: it just doesn't work"

tleydxdy replied:
indeed, after switching to docker-ce the postgres user now has a shell, rather than `/sbin/nologin` wtf. this doesn't make any sense. so different docker version will create different container from a image???

leonklingele commented:
Rebased & updated to `postgres:10-alpine`.

The upgrade was working fine for me this time, all data was transferred successfully..
@omarroth do you have a heavy database at hand to ensure an upgrade works there too?

doc75 commented:
On a fresh install, just changing the `docker-compsoe.yml` file from:
```yaml
postgres:
    image: postgres:10
```
to:
```yaml
postgres:
    image: postgres:10-alpine
```
is working fine.

Is there a real need to have code modification to support alpine version of postgres ?
It might be needed only for existing instance who want to migrate (not sure I have not tested this use-case)

Perflyst commented:
We don't build our own postgres image anymore. The postgres image can be adjusted in docker-compose.yml on initial deployment. Changing this for everyone who uses the compose file is not such a good idea as it could lead to data corruption.
